File: data-structures-sorting-algorithms-advanced/src/algorithms/bubbleSort.js
```javascript
/**
 * @fileoverview Implementation of the Bubble Sort algorithm.
 */

const { swap } = require('../utils/arrayUtils');

/**
 * Sorts an array of numbers using the Bubble Sort algorithm.
 *
 * Bubble Sort repeatedly steps through the list, compares adjacent elements
 * and swaps them if they are in the wrong order. The pass through the list
 * is repeated until no swaps are needed, which indicates that the list is sorted.
 *
 * Characteristics:
 * - Simple to understand and implement.
 * - Stable: Yes (maintains the relative order of equal elements).
 * - In-place: Yes (does not require extra memory beyond a few temporary variables).
 * - Adaptive: Yes (can be optimized to detect if array is already sorted and stop early).
 *
 * @param {Array<number>} arr The array of numbers to be sorted.
 * @returns {Array<number>} The sorted array. The original array is modified in-place.
 *
 * Time Complexity:
 *   - Worst-case: O(N^2) (e.g., reverse-sorted array)
 *   - Average-case: O(N^2)
 *   - Best-case: O(N) (if array is already sorted and optimization is used)
 *
 * Space Complexity:
 *   - O(1) (in-place)
 *
 * Example:
 *   bubbleSort([5, 1, 4, 2, 8]) // returns [1, 2, 4, 5, 8]
 *   bubbleSort([])             // returns []
 *   bubbleSort([1])            // returns [1]
 */
function bubbleSort(arr) {
  if (!Array.isArray(arr)) {
    throw new Error('Input must be an array.');
  }
  const n = arr.length;

  // An array with 0 or 1 element is already sorted.
  if (n <= 1) {
    return arr;
  }

  // Outer loop for passes through the array
  // After each pass, the largest unsorted element "bubbles up" to its correct position.
  for (let i = 0; i < n - 1; i++) {
    let swapped = false; // Flag to optimize: if no swaps occur in a pass, array is sorted.

    // Inner loop for comparisons and swaps
    // We only need to go up to n - 1 - i because the last `i` elements are already sorted.
    for (let j = 0; j < n - 1 - i; j++) {
      // Compare adjacent elements
      if (arr[j] > arr[j + 1]) {
        // Swap if they are in the wrong order
        swap(arr, j, j + 1);
        swapped = true;
      }
    }

    // If no two elements were swapped by inner loop, then the array is sorted
    if (!swapped) {
      break;
    }
  }
  return arr;
}

module.exports = bubbleSort;
```

File: data-structures-sorting-algorithms-enhanced/sorting_algorithms_project/algorithms/bubble_sort.py
def bubble_sort(arr):
    """
    Sorts an array using the Bubble Sort algorithm.

    Bubble Sort is a simple sorting algorithm that repeatedly steps through the list,
    compares adjacent elements and swaps them if they are in the wrong order.
    The pass through the list is repeated until no swaps are needed, which indicates
    that the list is sorted.

    Args:
        arr (list): The list of comparable elements to be sorted.

    Time Complexity:
        - Worst Case: O(n^2) - When the array is reverse sorted.
        - Average Case: O(n^2) - On average, it requires many comparisons and swaps.
        - Best Case: O(n) - When the array is already sorted (optimization for early exit).

    Space Complexity:
        - O(1) - It sorts the array in-place, requiring no additional storage
                 proportional to the input size.

    Stability:
        - Stable: It preserves the relative order of equal elements.
    """
    n = len(arr)
    # Traverse through all array elements
    for i in range(n - 1):
        # last_i elements are already in place
        # After each pass, the largest unsorted element "bubbles up" to its correct position
        swapped = False
        for j in range(0, n - i - 1):
            # Traverse the array from 0 to n-i-1
            # Swap if the element found is greater than the next element
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        # If no two elements were swapped by inner loop, then the array is sorted
        if not swapped:
            break

    # The function modifies the array in-place, so it doesn't explicitly return anything
    # but the array `arr` will be sorted after the call.
```